### PR TITLE
Accept shape tensors in Compiler

### PR DESCRIPTION
**Pull request by jhalakpatel**

### CODE REVIEW

File: tripy/tests/backend/test_compiler_api.py [hunk 1]
```diff
@@ -62,6 +62,14 @@ class TestInput:
             ([(1, 2, 3)], (1,), (2,), (3,)),
             # Only one value specified
             ([1], (1,), (1,), (1,)),
+            # one dynamic and one static dim
+            ([(1, 2, 3), 4], (1, 4), (2, 4), (3, 4)),
+            # Both dim dynamic
+            ([(1, 2, 3), (4, 5, 6)], (1, 4), (2, 5), (3, 6)),
+            # min/opt/max specified as shape tensor
+            ([tp.Shape([1, 2, 3])], (1,), (2,), (3,)),
```
Comment (jhalakpatel) [inline tripy/tests/backend/test_compiler_api.py:70]: @pranavm-nvidia this is how you expected shape profile to be provided using shape tensor?

File: tripy/tripy/backend/compiler_api.py [hunk 2]
```diff
@@ -80,20 +80,25 @@ def __init__(
...
-                if len(elem) != 3:
-                    raise_error(
-                        "Incorrect number of shape values provided.",
-                        [
-                            f"Exactly 3 shape values must be provided for each dimension (min/opt/max)"
-                            f" but got: {len(elem)} values in shape: {shape}. "
-                        ],
-                    )
+            elif isinstance(elem, Shape):
+                elem = elem.data().data()
```
Comment (jhalakpatel) [inline tripy/tripy/backend/compiler_api.py:84]: This is not suboptimal I guess. We should be able to slice shape tensors and populate shape bounds.

Comment (parthchadha) [reply tripy/tripy/backend/compiler_api.py:84]: Why is `elem` a shape tensor? What you want is the shape itself be a shape tensor describing the shape of input. 
```
a = tp.ones((2,3))
shape_of_a = a.shape
a_info = tp.InputInfo(shape_of_a, dtype=tp.float32) 
```
Your current approach is doing:
```
shape_of_a = a.shape
a_info = tp.InputInfo([shape_of_a[0], shape_of_a[1]], dtype=tp.float32) 
```

**Comment by jhalakpatel** [comment]
```
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================================================================== 1962 passed, 47 skipped, 1 deselected, 1 warning in 336.70s (0:05:36) ==================================================================================================
```

File: tripy/tripy/backend/compiler_api.py [hunk 3]
```diff
@@ -67,36 +69,44 @@ def __init__(
...
-                elem = (elem,) * 3
-            elif isinstance(elem, Sequence):
-                if not all(isinstance(val, numbers.Number) for val in elem):
-                    raise_error(
-                        "Shape values must be numbers.",
-                        [f"Shape: {shape} contains an element: {repr(elem)} with non-numerical value(s)"],
-                    )
-                if len(elem) != 3:
+        if isinstance(shape, Shape):
+            assert shape.shape.rank == 1
```
Comment (parthchadha) [inline tripy/tripy/backend/compiler_api.py:76]: Why is this `shape.shape.rank` and not `shape.rank`?

File: tripy/tests/backend/test_compiler_api.py [hunk 4]
```diff
@@ -95,7 +101,7 @@ def test_invalid_shape(self, shape, expected_error):
 @pytest.fixture(scope="session")
 def single_return_executable():
     compiler = tp.Compiler(add)
-    return compiler.compile(tp.InputInfo((2, 2), dtype=tp.float32), tp.InputInfo((2, 2), dtype=tp.float32))
+    return compiler.compile(tp.InputInfo(tp.Shape([2, 2]), dtype=tp.float32), tp.InputInfo((2, 2), dtype=tp.float32))
```
Comment (pranavm-nvidia) [inline tripy/tests/backend/test_compiler_api.py:104]: The idea was to support shape tensor inputs at runtime. I don't think we need to change anything about the compiler. `InputInfo` can be used to express the range of values the shape tensor can take.

File: tripy/tripy/backend/compiler_api.py [hunk 5]
```diff
@@ -67,36 +69,44 @@ def __init__(
...
-            if isinstance(elem, numbers.Number):
-                elem = (elem,) * 3
-            elif isinstance(elem, Sequence):
-                if not all(isinstance(val, numbers.Number) for val in elem):
-                    raise_error(
-                        "Shape values must be numbers.",
-                        [f"Shape: {shape} contains an element: {repr(elem)} with non-numerical value(s)"],
-                    )
-                if len(elem) != 3:
+        if isinstance(shape, Shape):
```
Comment (pranavm-nvidia) [inline tripy/tripy/backend/compiler_api.py:75]: I don't think there's any value in allowing shape tensors to `compile`. `InputInfo` already expresses everything we want AFAICT.

**Comment by jhalakpatel** [comment]
Closing this PR based on the above discussion.

**Pull request closed without merging**